frontend/catalyst/jax_primitives.py

```diff
@@ -1653,33 +1661,58 @@ def _hamiltonian_lowering(jax_ctx: mlir.LoweringRuleContext, coeffs: ir.Value, *
 #
 # sample measurement
 #
-@sample_p.def_abstract_eval
-def _sample_abstract_eval(obs, shots, shape):
-    assert isinstance(obs, AbstractObs)
-
+def sample_staging_rule(jaxpr_trace, obs, shots, num_qubits):
```

Comment:
Can you add documentation about this custom staging rule?

Reply:
done

Reply:
Can you also comment here how a custom staging rule is different from the abstract evaluation? I am just not grasping too well what the staging rule is doing.

Reply:
https://github.com/PennyLaneAI/catalyst/pull/1310/commits/e75f1ccc5bd9139ddd3b4fb507ca0ec30dc9118f I added links to jax source regarding why and how the custom staging rules changes the default primitive binding behavior